v2/service/module/watcher/watcher.go

```diff
@@ -0,0 +1,95 @@
+// Package watcher provides file update notification functionality as a module.
+package watcher
+
+import (
+	"fmt"
+
+	"github.com/fsnotify/fsnotify"
+)
+
+var (
+	ErrMissingWithFilename = fmt.Errorf("watcher.Watcher missing WithFilename option")
+	ErrMissingWithFunc     = fmt.Errorf("watcher.Watcher missing WithFunc option")
+)
+
+type Watcher struct {
+	w        *fsnotify.Watcher
+	stopped  chan struct{}
+	filename *string
+	fn       func() error
+	opts     []Opt
+}
+
+// New creates watcher with given options.
+// WithFilename and WithFunc options are mandatory.
+func New(opts ...Opt) *Watcher {
+	return &Watcher{opts: opts}
+}
+
+func (w *Watcher) Init() error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("watcher.Watcher error: %w", err)
+	}
+	w.w = watcher
+	w.stopped = make(chan struct{})
+
+	for _, opt := range w.opts {
+		if err := opt(w); err != nil {
+			return fmt.Errorf("watcher.Watcher Option error: %w", err)
+		}
+	}
+
+	switch {
+	case w.filename == nil:
+		return ErrMissingWithFilename
+	case w.fn == nil:
+		return ErrMissingWithFunc
+	}
+
+	return w.w.Add(*w.filename)
+}
+
+func (w *Watcher) Run() error {
+	for {
+		select {
+		case event := <-w.w.Events:
+			if event.Op&fsnotify.Write == fsnotify.Write {
+				err := w.fn()
+				if err != nil {
+					return err
+				}
+			}
+		case err := <-w.w.Errors:
+			return fmt.Errorf("watcher.Watcher error: %w", err)
+		case <-w.stopped:
+			return nil
```

Comment:
Based on the documentation on https://pkg.go.dev/github.com/fsnotify/fsnotify#Watcher.Close using `w.stopped` to signal closing isn't required. Instead we should do:
```go
select {
case event, ok := <-w.w.Events:
	if !ok {
		return nil
	}
```